File: vm/src/Execution/State.hs
{-
-- EPITECH PROJECT, 2025
-- Delivery
-- File description:
-- State
-}

module Execution.State (VMState(..), newVMState) where

import Types (Value, Instruction, FunctionMeta)
import qualified Data.Map as Map

-- === VM State ===
-- This struct represents the state of the virtual machine.

data VMState = VMState {
    stack     :: [Value],               -- The stack of the virtual machine. (Store values)
    env       :: Map.Map String Value,  -- The environment of the virtual machine. (Store variables)
    ip        :: Int,                   -- The instruction pointer. (Current instruction)
    callStack :: [(Int, [Value])],      -- The call stack. (Store return addresses and previous arguments)
    curArgs   :: [Value],               -- The current function arguments.
    isRunning :: Bool,                  -- The boolean to know if the virtual machine is running.

    -- === VM Data ===
    instructions :: [Instruction],      -- The instructions of the virtual machine. (Store instructions)
    constants :: [Value],               -- The constants of the virtual machine. (Store constants)
    functions :: [FunctionMeta]        -- The functions of the virtual machine. (Store functions)
} deriving (Show, Eq)

newVMState :: [Instruction] -> [Value] -> [FunctionMeta] -> VMState
newVMState insts consts funcs = VMState {
    stack = [],
    env = Map.empty,
    ip = 0,
    callStack = [],
    curArgs = [],
    isRunning = False,
    instructions = insts,
    constants = consts,
    functions = funcs
}

File: bonus/vm/src/execution/state.py
from dataclasses import dataclass, field
from typing import List, Dict, Optional, Tuple
from vm_types import Value, Instruction, FunctionMeta, IntVal, StringVal, ListVal

@dataclass
class VMState:
    stack: List[Value] = field(default_factory=list)
    env: Dict[str, Value] = field(default_factory=dict)
    ip: int = 0
    call_stack: List[Tuple[int, List[Value]]] = field(default_factory=list) # (return_ip, previous_args)
    cur_args: List[Value] = field(default_factory=list)
    is_running: bool = False
    
    instructions: List[Instruction] = field(default_factory=list)
    constants: List[Value] = field(default_factory=list)
    functions: List[FunctionMeta] = field(default_factory=list)

def new_vm_state(insts: List[Instruction], consts: List[Value], funcs: List[FunctionMeta], args: List[str]) -> VMState:
    initial_env = {
        "argc": IntVal(len(args)),
        "argv": ListVal([StringVal(arg) for arg in args])
    }
    
    return VMState(
        stack=[],
        env=initial_env,
        ip=0,
        call_stack=[],
        cur_args=[],
        is_running=False,
        instructions=insts,
        constants=consts,
        functions=funcs
    )
